File: apps/ios/Voiid/Voiid/Main/Settings/LinkBrowserCode.swift
import Foundation

/// Accept only Voiid's companion QR payload. Never follow an arbitrary scanned URL.
enum LinkBrowserCode {
    static func token(from raw: String) -> String? {
        guard raw.utf8.count < 256,
              let url = URLComponents(string: raw),
              url.scheme == "voiid", url.host == "link",
              url.user == nil, url.password == nil, url.port == nil,
              url.path.isEmpty, url.fragment == nil,
              let items = url.queryItems, items.count == 1,
              items[0].name == "token", let token = items[0].value, token.utf8.count == 32,
              token.range(of: "^[A-Za-z0-9_-]{32}$", options: .regularExpression) != nil
        else { return nil }
        return token
    }
}

File: apps/android/app/src/main/java/com/voiid/app/net/LinkBrowserCode.kt
package com.voiid.app.net

import java.net.URI

/**
 * Accept ONLY Voiid's companion QR payload — `voiid://link?token=<32 url-safe chars>` — and
 * never follow an arbitrary scanned URL. Twin of iOS `LinkBrowserCode.swift`, same rules.
 */
object LinkBrowserCode {
    private val TOKEN = Regex("^[A-Za-z0-9_-]{32}$")

    fun token(raw: String): String? {
        if (raw.toByteArray().size >= 256) return null
        val uri = runCatching { URI(raw) }.getOrNull() ?: return null
        if (uri.scheme != "voiid" || uri.host != "link") return null
        if (uri.rawUserInfo != null || uri.port != -1 || uri.rawFragment != null) return null
        if (!uri.rawPath.isNullOrEmpty()) return null
        val items = uri.rawQuery?.split("&") ?: return null
        if (items.size != 1) return null
        val (name, value) = items[0].split("=", limit = 2).takeIf { it.size == 2 } ?: return null
        if (name != "token" || !TOKEN.matches(value)) return null
        return value
    }
}
